cedar-policy-validator/src/custom_schema/ast.rs

```diff
@@ -0,0 +1,299 @@
+use cedar_policy_core::{
+    ast::Id,
+    parser::{
+        cst::{Name, Ref},
+        Loc, Node,
+    },
+};
+use itertools::Either;
+use itertools::Itertools;
+use nonempty::NonEmpty;
+use smol_str::SmolStr;
+
+use crate::SchemaTypeVariant;
+
+const IPADDR_EXTENSION: &str = "ipaddr";
+const DECIMAL_EXTENSION: &str = "decimal";
+pub const EXTENSIONS: [&str; 2] = [IPADDR_EXTENSION, DECIMAL_EXTENSION];
+pub const BUILTIN_TYPES: [&str; 3] = ["Long", "String", "Bool"];
+
+pub(super) const CEDAR_NAMESPACE: &str = "__cedar";
+
+pub type IdentOrString = Either<Node<Id>, Node<SmolStr>>;
+pub type Schema = Vec<Node<Namespace>>;
+
+/// A path is a non empty list of identifiers that forms a namespace + type
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Path(pub Node<NonEmpty<Id>>);
+
+impl Path {
+    /// Create a [`Path`] with a single entry
+    pub fn single(node: Id, loc: Loc) -> Self {
+        Self(Node::with_source_loc(NonEmpty::new(node), loc))
+    }
+
+    /// Create [`Path`] with a head and an iterator
+    pub fn new(head: Id, rest: impl IntoIterator<Item = Id>, loc: Loc) -> Self {
```

Comment:
for the doc comment: is `head` the first component of the path, or the last component (the actual typename)?